### PR TITLE
sources: Support multiple authentication types

### PR DESCRIPTION
**Pull request by rvsia**
Modification according to https://github.com/ManageIQ/sources-api/pull/97

**Description**

- adds support for multiple authentication
- adds support for more advanced schemas in DB

# Format for sourceType schema

```rb
schema = {
  :authentication => [{
    :type             => 'username_password', # same as in fields - authtype
    :name             => 'Token', # Will be shown in the authentication selection step
    :fields           => [
      # DDF Fields, will be shown in the authentication selection step
      {:component => "text-field", :name => "authentication.authtype", :type => "hidden", :initialValue => "username_password"}
    ],
    :additional_steps => [
      # Steps according to DDF
      # Don't use `stepKey` in the first step
      # Don't use `nextStep` in the last step
      # => These attributes will be appended automatically
      # Additional steps are shown after the authentication_type selection step
    ]
  }],
  :endpoint       => {
    :title  => 'Title', # If the endpoint will be on a separate step
    :hidden => true, # If true, the endpoint step will be hidden, use if all components have `:hideField => true`
    :fields => [
      # DDF Fields
      {:component => "checkbox", :name => "endpoint.verify_ssl", :label => "Verify SSL"}
    ]
  }
}
```

TODO:

- [x] endpoint step
- [x] auth selector
- [x] additional steps
- [x] modify summary
- [x] tests
- [x] docs

finalfinalfinal

**HOW TO TEST IT WITHOUT THE API?**

Easy peasy lemon squeezy :tangerine: 

1. Link it with Sources-UI
2. use content of ` packages/sources/src/tests/helpers/sourceTypes.js` as a return value for redux action calling `doLoadSourceTypes` :

```
export const loadSourceTypes = () => (dispatch) => {
    dispatch({ type: ACTION_TYPES.LOAD_SOURCE_TYPES_PENDING });

    return doLoadSourceTypes().then(sourceTypes => dispatch({
        type: ACTION_TYPES.LOAD_SOURCE_TYPES_FULFILLED,
        payload: SOURCE_TYPES // <------------------- HERE
    }));
};
```

### CODE REVIEW

**Comment by rvsia** [comment]
@Hyperkid123 Please, review :pray:

File: packages/sources/src/tests/helpers/sourceTypes.js [hunk 1]
```diff
@@ -1,39 +1,298 @@
...
-                component: 'text-field',
-                name: 'field'
-            }]
-        }
+            authentication: [{
+                type: 'token',
+                name: 'Token',
+                fields: [
+                    {
+                        component: 'text-field',
```
[the remark this answers is not in the document]
Comment (rvsia) [reply packages/sources/src/tests/helpers/sourceTypes.js]: This is a basically an 'offline' API response and API returns strings.

File: packages/sources/src/tests/sourceFormRenderer/components/authSelect.test.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            sourceTypes,
+            FieldProvider,
+            input: {
+                name: 'auth-select',
+                onChange: spyOnChange
+            },
+            formOptions: {
+                getState: () => ({
+                    values: {
+                        source_type: 'amazon' //eslint-disable-line camelcase
```
[the remark this answers is not in the document]
Comment (rvsia) [reply packages/sources/src/tests/sourceFormRenderer/components/authSelect.test.js]: I agree @karelhala

File: packages/sources/src/tests/helpers/sourceTypes.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hyperkid123) [reply packages/sources/src/tests/helpers/sourceTypes.js]: Oh i see just a mock. :ok:

File: packages/sources/src/tests/sourceFormRenderer/components/authSelect.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hyperkid123) [reply packages/sources/src/tests/sourceFormRenderer/components/authSelect.test.js]: @rvsia  if not for the whole repo i think adding little eslint config just for sources should be ok.

**Comment by Hyperkid123** [comment]
@rvsia ping me when ever the API is ready. :red_circle:

**Comment by rvsia** [comment]
@Hyperkid123 It is ready, after releasing this I will finish the Sources-UI update PR (https://github.com/ManageIQ/sources-ui/pull/111) and then, the UI and the API PRs could be merged. :dagger: :dancer:

File: packages/sources/.eslintrc.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+    "rules": {
+        "camelcase": "off"
+    }
+}
```
Comment (Hyperkid123) [inline packages/sources/.eslintrc.json]: New line :eyes:

**Pull request merged**